geomesa-convert/geomesa-convert-common/src/main/scala/org/locationtech/geomesa/convert/SimpleFeatureValidator.scala

```diff
@@ -8,168 +8,256 @@
+    * @param error error message
     */
-  class ZIndexValidator extends SimpleFeatureValidator {
-
-    import org.locationtech.geomesa.utils.geotools.RichSimpleFeatureType.RichSimpleFeatureType
-    import org.locationtech.geomesa.utils.geotools.wholeWorldEnvelope
-
-    private var dtgIndex: Int = -1
-    private var geomIndex: Int = -1
-    private var minDate: Date = _
-    private var maxDate: Date = _
-    private var lastErr: String = _
-
-    override def name: String = ZIndexValidator.name
-
-    override def init(sft: SimpleFeatureType): Unit = {
-      lastErr = null
-      dtgIndex = sft.getDtgIndex.getOrElse(-1)
-      geomIndex = sft.getGeomIndex
-      minDate = Date.from(BinnedTime.ZMinDate.toInstant)
-      maxDate = Date.from(BinnedTime.maxDate(sft.getZ3Interval).toInstant)
+  private class NullValidator(i: Int, error: String) extends Validator {
+    override def validate(sf: SimpleFeature): String = if (sf.getAttribute(i) == null) { error } else { null}
+  }
 
-      if (dtgIndex == -1 || geomIndex == -1) {
-        throw new IllegalArgumentException("Z3 validator cannot be used on a type lacking a date or geometry")
+  /**
+    * Z2 validator
+    *
+    * @param geom geom index
+    */
+  private class Z2Validator(geom: Int) extends Validator {
+    override def validate(sf: SimpleFeature): String = {
+      val g = sf.getAttribute(geom).asInstanceOf[Geometry]
+      if (g == null) {
+        Errors.GeomNull
+      } else if (!wholeWorldEnvelope.contains(g.getEnvelopeInternal)) {
+        Errors.GeomBounds
+      } else {
+        null
       }
     }
+  }
 
-    override def validate(sf: SimpleFeature): Boolean = {
-      lastErr = null
-      val date = sf.getAttribute(dtgIndex).asInstanceOf[Date]
-      val geom = sf.getAttribute(geomIndex).asInstanceOf[Geometry]
-      if (date == null) {
-        lastErr = "Null date attribute"
-        false
-      } else if (geom == null) {
-        lastErr = "Null geometry attribute"
-        false
-      } else if (date.before(minDate)) {
-        lastErr = s"Date is before Z3 Min Date ($minDate)"
-        false
-      } else if (date.after(maxDate)) {
-        lastErr = s"Date is after Z3 Max Date ($maxDate)"
-        false
-      } else if (!wholeWorldEnvelope.contains(geom.getEnvelopeInternal)) {
-        lastErr = s"Geometry exceeds world bounds ($wholeWorldEnvelope)"
-        false
+  /**
+    * Z3 validator
+    *
+    * @param geom geom index
+    * @param dtg dtg index
+    * @param minDate min z3 date
+    * @param maxDate max z3 date
+    */
+  private class Z3Validator(geom: Int, dtg: Int, minDate: Date, maxDate: Date) extends Validator {
+    private val dateBefore = s"date is before minimum indexable date ($minDate)"
+    private val dateAfter = s"date is after maximum indexable date ($maxDate)"
+
+    override def validate(sf: SimpleFeature): String = {
+      val d = sf.getAttribute(dtg).asInstanceOf[Date]
+      val g = sf.getAttribute(geom).asInstanceOf[Geometry]
```

Comment:
it could definitely be more compact - i was going for minimal processing. i was trying to figure out a way to avoid evaluating anything unnecessarily and and this was all i could come up with